lib/posix/eventfd.c

```diff
@@ -246,51 +262,181 @@ static const struct fd_op_vtable eventfd_fd_vtable = {
+
+	/* do not hold a spinlock when taking a mutex */
+	k_spin_unlock(&efd->lock, key);
+	err = k_mutex_lock(lock, K_FOREVER);
+	__ASSERT(err == 0, "k_mutex_lock() failed: %d", err);
 
-	for (i = 0; i < ARRAY_SIZE(efds); ++i) {
-		if (!(efds[i].flags & EFD_IN_USE)) {
-			efd = &efds[i];
+	while (true) {
+		/* retake the spinlock */
+		key = k_spin_lock(&efd->lock);
+
+		ret = op(efd, buf);
+		switch (ret) {
+		case -EAGAIN:
+			/* not an error in blocking mode. break and try again */
 			break;
+		case 0:
+			/* success! */
+			ret = sizeof(eventfd_t);
+			goto unlock_mutex;
+		default:
+			/* some other error */
+			__ASSERT_NO_MSG(ret < 0);
+			errno = -ret;
+			ret = -1;
+			goto unlock_mutex;
 		}
+
+		/* do not hold a spinlock when taking a mutex */
+		k_spin_unlock(&efd->lock, key);
+
+		/* wait for a write or close */
+		err = k_condvar_wait(cond, lock, K_FOREVER);
+		__ASSERT(err == 0, "k_condvar_wait() failed: %d", err);
 	}
 
-	if (efd == NULL) {
+unlock_mutex:
+	k_spin_unlock(&efd->lock, key);
+	/* only wake a single waiter */
+	err = k_condvar_signal(cond);
+	__ASSERT(err == 0, "k_condvar_signal() failed: %d", err);
+	err = k_mutex_unlock(lock);
+	__ASSERT(err == 0, "k_mutex_unlock() failed: %d", err);
+	goto out;
+
+unlock_spin:
+	k_spin_unlock(&efd->lock, key);
+
+out:
+	return ret;
+}
+
+/*
+ * Public-facing API
+ */
+
+int eventfd(unsigned int initval, int flags)
+{
+	int fd = 1;
+	size_t offset;
+	struct eventfd *efd = NULL;
+
+	if (flags & ~EFD_FLAGS_SET_INTERNAL) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (sys_bitarray_alloc(&efds_bitarray, 1, &offset) < 0) {
 		errno = ENOMEM;
-		goto exit_mtx;
+		return -1;
 	}
 
+	efd = &efds[offset];
+
 	fd = z_reserve_fd();
 	if (fd < 0) {
-		goto exit_mtx;
+		sys_bitarray_free(&efds_bitarray, 1, offset);
+		return -1;
 	}
 
-	efd->flags = EFD_IN_USE | flags;
+	efd->flags = EFD_IN_USE_INTERNAL | flags;
 	efd->cnt = initval;
 
-	k_poll_signal_init(&efd->write_sig);
-	k_poll_signal_init(&efd->read_sig);
-	z_waitq_init(&efd->wait_q);
+	z_finalize_fd(fd, efd, &eventfd_fd_vtable);
 
-	if (initval != 0) {
-		k_poll_signal_raise(&efd->read_sig, 0);
+	return fd;
+}
+
+int eventfd_read(int fd, eventfd_t *value)
+{
+	int ret;
+	void *obj;
+
+	obj = z_get_fd_obj(fd, &eventfd_fd_vtable, EBADF);
+	if (obj == NULL) {
+		return -1;
 	}
-	k_poll_signal_raise(&efd->write_sig, 0);
 
-	z_finalize_fd(fd, efd, &eventfd_fd_vtable);
+	ret = eventfd_rw_op(obj, value, sizeof(eventfd_t), eventfd_read_locked);
+	__ASSERT_NO_MSG(ret == -1 || ret == sizeof(eventfd_t));
+	if (ret < 0) {
+		return -1;
+	}
 
-exit_mtx:
-	k_mutex_unlock(&eventfd_mtx);
-	return fd;
+	return 0;
+}
+
+int eventfd_write(int fd, eventfd_t value)
+{
+	int ret;
+	void *obj;
+
+	obj = z_get_fd_obj(fd, &eventfd_fd_vtable, EBADF);
+	if (obj == NULL) {
+		return -1;
+	}
+
+	ret = eventfd_rw_op(obj, &value, sizeof(eventfd_t), eventfd_write_locked);
+	__ASSERT_NO_MSG(ret == -1 || ret == sizeof(eventfd_t));
+	if (ret < 0) {
```

Comment:
Is this superfluous? Does it have to be -1 can you just return ret?

Reply:
The return value needs to be either 0 or -1, but `ret` will be either -1 or `sizeof(eventfd_t)`.

Technically, I could have made a comparison against 0 as well, I just tend to prefer branching in the error case.